libdnf/comps/environment/environment-private.hpp

```diff
@@ -0,0 +1,47 @@
+/*
+Copyright Contributors to the libdnf project.
+
+This file is part of libdnf: https://github.com/rpm-software-management/libdnf/
+
+Libdnf is free software: you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation, either version 2.1 of the License, or
+(at your option) any later version.
+
+Libdnf is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+
+#ifndef LIBDNF_COMPS_ENVIRONMENT_ENVIRONMENT_PRIVATE_HPP
+#define LIBDNF_COMPS_ENVIRONMENT_ENVIRONMENT_PRIVATE_HPP
+
+
+#include <libdnf/comps/environment/environment.hpp>
+
+extern "C" {
+#include <solv/pool.h>
+}
+
+namespace libdnf::comps {
+
+
+void add_solvable_id(Environment & environment, Id solvable_id);
+
+
+void add_solvable_ids(Environment & environment, std::vector<Id> solvable_ids);
+
+
+// Search solvables that correspond to the environment_ids for given key
+// Return first non-empty string
+std::string lookup_str(Pool * pool, std::vector<EnvironmentId> environment_ids, Id key);
```

Comment:
Have a look at `libdnf::solv::Pool::lookup_str()`, should do what you need.

Reply:
(I haven't said it explicitly but it seems you can drop this header altogether, doesn't seem to be needed)

Reply:
The ``libdnf::solv::Pool::lookup_str()`` returns the string only for one id, but I need to search over all the environment ids.

Reply:
Okay, sorry haven't noticed. You could still use the `libdnf::solv::Pool::lookup_str()` in a loop, and since you need that in 3 places in `environment.cpp`, you could put that into a local function. I see the same method is also duplicated in `comps/group.cpp`, so perhaps we should share that function more globally?

I'm not entirely sure what's the situation, are the `environment_ids` multiple libsolv `Id`s for environments with the same string id from multiple repos? And are they somehow guaranteed to e.g. have the same name or could it happen you could get different names depending on which one is first in the `environment_ids` vector?

Reply:
> Okay, sorry haven't noticed. You could still use the `libdnf::solv::Pool::lookup_str()` in a loop, and since you need that in 3 places in `environment.cpp`, you could put that into a local function. I see the same method is also duplicated in `comps/group.cpp`, so perhaps we should share that function more globally?

It might make sense, but I am not sure how to do it when groups have ``GroupId`` objects and environments have ``EnvironmentId`` and there is no duck-typing. But it's both just for storing solvable id, so maybe it can be united.

> I'm not entirely sure what's the situation, are the `environment_ids` multiple libsolv `Id`s for environments with the same string id from multiple repos? And are they somehow guaranteed to e.g. have the same name or could it happen you could get different names depending on which one is first in the `environment_ids` vector?

Yes, they are for environments with the same string id. They have to be merged and it can be different for each query (because different repos can be enabled). There were two options:
1. Create one merged environment object when query is created
2. Store only the solvable ids and iterate through them when merged attribute is needed

Originally, we've decided to use the option 2 because I hadn't found a way to iterate over all translations and could only return translation when language was given. This has changed (it's used in ``libdnf::comps::Environment::dump``), but the option 2 stayed.

What is still missing is to order the solvables, because it hasn't been yet decided how it should be ordered.

Reply:
> It might make sense, but I am not sure how to do it when groups have `GroupId` objects and environments have `EnvironmentId` and there is no duck-typing. But it's both just for storing solvable id, so maybe it can be united.

Oh, I see. Okay just leave it this way, but you don't need the `environment-private.hpp` to share the function with other modules, right?

And we can leave the merging topic for later, but thanks for the explanation :slightly_smiling_face: